File: 3583-count-special-triplets/3583-count-special-triplets.cpp
class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        unordered_map<int,int> left,right;
        long ans = 0;
        for(int i=1;i<nums.size();++i) ++right[nums[i]]; ++left[nums[0]];
        for(int i=1;i<nums.size()-1;++i){
            --right[nums[i]];
            if(left.count(nums[i]*2) && right.count(nums[i]*2)){
                ans += ((long)left[nums[i]*2])*((long)right[nums[i]*2]);
                ans %= 1000000007;
            }
            ++left[nums[i]];
        }
        return ans;
    }
};

File: 3583-count-special-triplets/3583-count-special-triplets.java
class Solution {
    public int specialTriplets(int[] nums) {
        int n = nums.length;
        long ans = 0;
        int mod = 1000000007;

        HashMap<Integer,Integer> left = new HashMap<>();
        HashMap<Integer,Integer> right = new HashMap<>();

        for(int i=2;i<n;++i) right.put(nums[i], right.getOrDefault(nums[i],0) + 1);

        left.put(nums[0],1);

        for(int i=1;i<n-1;++i){
            int l = left.getOrDefault(nums[i]*2,0);
            int r = right.getOrDefault(nums[i]*2,0);
            ans = (ans + ((long)l * (long)r))%mod;
            
            left.put(nums[i],left.getOrDefault(nums[i],0) + 1);
            right.put(nums[i+1], right.get(nums[i+1]) - 1);
        }

        return (int)ans;
    }
}